impl/src/main/java/com/okta/sdk/impl/util/OAuth2Utils.java

```diff
@@ -0,0 +1,163 @@
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
+
+public class OAuth2Utils {
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    private static List<String> allowedSigningAlgorithms =
+        new ArrayList<>(Arrays.asList("RS256", "RS384", "RS512", "ES256", "ES384", "ES512"));
+
+    private static final OkHttpRequestExecutorFactory okHttpRequestExecutorFactory =
+        new OkHttpRequestExecutorFactory();
+
+    private static KeyFactory getKeyFactoryInstance() throws NoSuchAlgorithmException {
+        return KeyFactory.getInstance("RSA");
+    }
+
+    public static String createSignedJWT(ClientConfiguration clientConfiguration)
+        throws RuntimeException, InvalidKeySpecException, NoSuchAlgorithmException, IOException {
+        String clientId = clientConfiguration.getClientId();
+        SignatureAlgorithm signatureAlgorithm = getSignatureAlgorithm(clientConfiguration.getAlgorithm());
+        PrivateKey privateKey = createPrivateKeyFromPemFilePath(clientConfiguration.getPrivateKey());
+        Instant now = Instant.now();
+
+        String jwt = Jwts.builder()
+            .setAudience(clientConfiguration.getBaseUrl() + "/oauth2/default/v1/token")
+            .setIssuedAt(Date.from(now))
+            .setExpiration(Date.from(now.plus(1L, ChronoUnit.HOURS)))
+            .setIssuer(clientId)
+            .setSubject(clientId)
+            .claim("jti", UUID.randomUUID().toString())
+            .signWith(signatureAlgorithm, privateKey)
+            .compact();
+
+        return jwt;
+    }
+
+    public static String getOAuth2AccessToken(ClientConfiguration clientConfiguration)
+        throws HttpException, IOException, InvalidKeySpecException, NoSuchAlgorithmException {
+        String signedJwt = createSignedJWT(clientConfiguration);
+
+        RequestExecutor requestExecutor = okHttpRequestExecutorFactory.create(clientConfiguration);
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.add("Accept", "application/json");
+        httpHeaders.add("Content-Type", "application/x-www-form-urlencoded");
+
+        QueryString queryString = new QueryString();
+        queryString.put("grant_type", "client_credentials");
+        queryString.put("scope", clientConfiguration.getScopes());
+        queryString.put("client_assertion_type",
+            URLEncoder.encode("urn:ietf:params:oauth:client-assertion-type:jwt-bearer", "UTF-8"));
+        queryString.put("client_assertion", signedJwt);
+
+        DefaultRequest accessTokenRequest =
+            new DefaultRequest(HttpMethod.POST, "/oauth2/v1/token", queryString, httpHeaders);
```

Comment:
Question to reviewers: Is the URI correct? or should it be _/oauth2/**default**/v1/token_ as in line #51 ?

Reply:
`/oauth2/v1/token` is correct, `/oauth2/<id>/v1/token` would be used for custom authorization servers which do NOT support the Okta scopes

Reply:
ok fine

Reply:
For the "scope" of this requirement, we would work only with okta scopes (from the requirements spec). So this should be fine, I think.